### ISSUE TITLE
Add hooks to cater for platform incompatibilities and UX tailoring

### ISSUE THREAD

**Issue by acicovic** [Parsely/wp-parsely#1338]
## Is your feature request related to a problem?
Our plugin is being used in platforms where:
- Some of our features may be incompatible, causing issues or breakages.
- The user experience (UX) may need to be tailored.

## Describe the solution you'd like
- Document which features can cause issues/breakage for the platforms in question.
- Introduce hooks so the specific features can be disabled.
- Document how user experience must be tailored for every platform.
- Introduce hooks that will allow this custom experience to happen.

**Comment by acicovic** [comment]
For the platforms and users that need to disable the Content Helper in whole or in part, here are the hooks I propose:

- `enable_content_helper`: When `false`, disables all PCH features.
- `enable_content_helper_dashboard_widget`: When `false`, disables the PCH Dashboard Widget.
- `enable_content_helper_post_list_column`: When `false`, disables the PCH column in Post/CPT listings.
- `enable_content_helper_editor_sidebar`: When `false`, disables the PCH Sidebar within Gutenberg.

All the above will be `true` by default.

@mehmoodak, @mjangda: ideas, comments, objections?

**Comment by mjangda** [comment]
Maybe we should change `enable_content_helper_sidebar` to `enable_content_helper_editor_sidebar` instead? Specifically, including `editor` provides some context around where the sidebar is, similar to the other two features (dashboard, post_list).

**Comment by acicovic** [comment]
Yes, good idea. I've updated my first post with this suggestion.

**Comment by mehmoodak** [comment]
> Some of our features may be incompatible, causing issues or breakages.

Although the hooks looks good to me but can you please explain a bit why we needed these hooks ..... is it just a safe guard if some feature is causing breakage or there is some other reason for it?

**Comment by acicovic** [comment]
> please explain a bit why we needed these hooks

The main reasons would be:
- I've heard of at least one Parse.ly customer who would want to disable the PCH (I'm unsure partly or totally).
- When the plugin is used in platforms that are not self-hosted installations, there may be incompatibilities that make the plugin crash, the plaform may want to tailor the experience of the plugin, or it may want to disable the features in certain sites.

**Comment by GaryJones** [comment]
Also, please [use prefixes in the hooks name](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#namespace-declarations), since these are global in the WP application. The plugin currently [supports two different prefixes](https://github.com/Parsely/wp-parsely/blob/e8b61231e05b632b31c67c523d6c634ff92559ec/.phpcs.xml.dist#L53-L54), so you'll need to see which is the most suitable / prevalent.

**Comment by acicovic** [comment]
This issue will need a couple of PRs to get closed and I'm currently working on one of them, still in draft. I'm hoping to open it for review tomorrow.

**Comment by acicovic** [comment]
We'll be closing this issue after #1724 gets merged. A separate issue will be opened if/when we need to extend this effort into additional platforms.
